include/libhal-soft/i2c_bit_bang.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+#include <libhal/i2c.hpp>
+#include <libhal/output_pin.hpp>
+#include <libhal/units.hpp>
+
+namespace hal {
+/**
+ * @brief A bit bang implementation for i2c. This implementation of i2c only
+ * needs 2 gpio's to work correctly. In its current implementation we only
+ * support single controller multiple peripheral. But have intentions of
+ * supporting multiple controller soon. Additionally, it's important to note
+ * that bit bang is a best try implementation of i2c. As in it will do it's best
+ * to reach the requested frequency but will typically be slower then requested
+ * with brief testing it seems that the error is about 2us per period for the
+ * lpc4078 , so you can try to account for this error with the frequency you set
+ * for example, achieving the correct period for 100kHz would result in setting
+ * slightly more then a 125kHz frequency. Additionally, the fastest frequency
+ * this bit bang can handle is around 200kHz.
```

Comment:
Can you fix the grammar mistakes. Its also a big confusing. The term you are looking for is "Best effort" and you can just say that the clock rate is best effort and is subject to the speed of the microcontroller as well as the steady clock provided.

You can explain that with the lpc4078 the maximum achieved clock rate was around 180kHz. It doesn't make sense to say that it IS 200kHz because a faster processor could do this faster. 

We are actually missing one additional bit of information. Interrupts disrupt this controller. Most devices should be able to support this, but some devices may not support the clock and data lines freezing for an arbitrary amount of time.